File: RLSilk/app/src/main/java/c/geonet/rlsilk/custom/Player.java
package c.geonet.rlsilk.custom;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

/**
 * Created by Ravi on 15.03.2018.
 */

public class Player {
    private String id;
    private String name;
    private int platform;
    private String avatarUrl;
    private String profileUrl;
    private int wins;
    private int goals;
    private int mvps;
    private int saves;
    private int shots;
    private int assists;
    private ArrayList<Season> seasons;
    private long updated;
    private long nextUpdate;

    public Season currentSeason(){
        return seasons.get(seasons.size()-1);
    }

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public int getPlatform() {
        return platform;
    }

    public String getAvatarUrl() {
        return avatarUrl;
    }

    public String getProfileUrl() {
        return profileUrl;
    }

    public int getWins() {
        return wins;
    }

    public int getGoals() {
        return goals;
    }

    public int getMvps() {
        return mvps;
    }

    public int getSaves() {
        return saves;
    }

    public int getShots() {
        return shots;
    }

    public int getAssists() {
        return assists;
    }

    public ArrayList<Season> getSeasons() {
        return seasons;
    }

    public long getUpdated() {
        return updated;
    }

    public long getNextUpdate() {
        return nextUpdate;
    }

    public void setId(String id) {
        this.id = id;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setPlatform(int platform) {
        this.platform = platform;
    }

    public void setAvatarUrl(String avatarUrl) {
        this.avatarUrl = avatarUrl;
    }

    public void setProfileUrl(String profileUrl) {
        this.profileUrl = profileUrl;
    }

    public void setWins(int wins) {
        this.wins = wins;
    }

    public void setGoals(int goals) {
        this.goals = goals;
    }

    public void setMvps(int mvps) {
        this.mvps = mvps;
    }

    public void setSaves(int saves) {
        this.saves = saves;
    }

    public void setShots(int shots) {
        this.shots = shots;
    }

    public void setAssists(int assists) {
        this.assists = assists;
    }

    public void setSeasons(ArrayList<Season> seasons) {
        this.seasons = seasons;
    }

    public void setUpdated(long updated) {
        this.updated = updated;
    }

    public void setNextUpdate(long nextUpdate) {
        this.nextUpdate = nextUpdate;
    }

    public static Player fromJSON(JSONObject response){
        Player player = new  Player();
        try {
            if (response.has("displayName"))
                player.setName(response.getString("displayName"));
            if (response.has("avatar"))
                player.setAvatarUrl(response.getString("avatar"));
            if (response.has("updatedAt"))
                player.setUpdated(response.getLong("updatedAt"));
            if (response.has("nextUpdateAt"))
                player.setNextUpdate(response.getLong("nextUpdateAt"));
            if (response.has("stats")) {
                JSONObject stats = response.getJSONObject("stats");
                player.setWins(stats.getInt("wins"));
                player.setGoals(stats.getInt("goals"));
                player.setMvps(stats.getInt("mvps"));
                player.setSaves(stats.getInt("saves"));
                player.setShots(stats.getInt("shots"));
                player.setAssists(stats.getInt("assists"));
            }
            if (response.has("profileUrl"))
                player.setProfileUrl(response.getString("profileUrl"));
            if (response.has("rankedSeasons")) {
                JSONObject season = response.getJSONObject("rankedSeasons");
                int i = 0;
                while (season.has("" + i + 1))
                    i++;
                season = season.getJSONObject("" + i);
                JSONObject duel = season.getJSONObject("10");
                JSONObject duo = season.getJSONObject("11");
                JSONObject solo = season.getJSONObject("12");
                JSONObject standard = season.getJSONObject("13");
                player.currentSeason().setStats(
                        new Ranking(solo.getInt("rankedPoints"), solo.getInt("matchesPlayed"), solo.getInt("tier"), solo.getInt("division"))
                        , new Ranking(duel.getInt("rankedPoints"), duel.getInt("matchesPlayed"), duel.getInt("tier"), duel.getInt("division"))
                        , new Ranking(duo.getInt("rankedPoints"), duo.getInt("matchesPlayed"), duo.getInt("tier"), duo.getInt("division"))
                        , new Ranking(standard.getInt("rankedPoints"), standard.getInt("matchesPlayed"), standard.getInt("tier"), standard.getInt("division"))
                        , player.getUpdated());
            }
        }catch (JSONException error){

        }finally{
            return player;
        }
    }

    public Player mergeWithPlayer(Player recent){
        for(int i = 0; i< recent.getSeasons().size();i++) {
            recent.getSeasons().get(i).getSolo().putAll(this.seasons.get(i).getSolo());
        }
        return recent;
    }
}

File: RLComp/app/src/main/java/com/example/ravi/rlcomp/custom/Player.kt
package com.example.ravi.rlcomp.custom

import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList

class Player {
    var id: String? = null
    var name: String? = null
    var platform: Int = 0
    var avatarUrl: String? = null
    var profileUrl: String? = null
    var wins: Int = 0
    var goals: Int = 0
    var mvps: Int = 0
    var saves: Int = 0
    var shots: Int = 0
    var assists: Int = 0
    var seasons: LinkedHashMap<Int,TimestampList> = LinkedHashMap()
    var updated: Long = 0
    var nextUpdate: Long = 0
    var currentSeason: Int = 0

    fun toText():String{
        var bob:StringBuilder = StringBuilder()
        bob.append(""" name: $name" wins: ${wins}goals: ${goals}mvps: ${mvps}saves: ${saves}shots: ${shots}assists: $assists""")
        for(key:Int in seasons.keys){
            bob.append(" Season: ")
            bob.append(key)
            bob.append("    ")
            bob.append(seasons[key]?.getGoals())
        }
        return bob.toString()

    }
    fun update(playerson : JSONObject){
        try {
            if (playerson.has("displayName"))
                this.name = playerson.getString("displayName")
            if (playerson.has("avatar"))
                this.avatarUrl = playerson.getString("avatar")
            if (playerson.has("updatedAt"))
                this.updated = playerson.getLong("updatedAt")
            if (playerson.has("nextUpdateAt"))
                this.nextUpdate = playerson.getLong("nextUpdateAt")
            if (playerson.has("stats")) {
                val stats = playerson.getJSONObject("stats")
                this.wins = stats.getInt("wins")
                this.goals = stats.getInt("goals")
                this.mvps = stats.getInt("mvps")
                this.saves = stats.getInt("saves")
                this.shots = stats.getInt("shots")
                this.assists = stats.getInt("assists")
            }
            if (playerson.has("profileUrl"))
                this.profileUrl = playerson.getString("profileUrl")
            if (playerson.has("rankedSeasons")) {
                var allSeasons = playerson.getJSONObject("rankedSeasons")
                for (key: String in allSeasons.keys()) {
                    var processedSeason = allSeasons.getJSONObject(key)
                    if (key.toInt() >= currentSeason) {
                        var matchesPlayed = processedSeason.getJSONObject("" + 10).getInt("matchesPlayed") + processedSeason.getJSONObject("" + 11).getInt("matchesPlayed")
                        +processedSeason.getJSONObject("" + 12).getInt("matchesPlayed") + processedSeason.getJSONObject("" + 13).getInt("matchesPlayed")
                        var rankings = getRankings(processedSeason)
                        if (key.toInt() > currentSeason)
                            seasons[key.toInt()] = TimestampList(Timestamp(this.updated, matchesPlayed, rankings, shots, goals,
                                    saves, assists,wins, mvps))
                        else {
                            seasons[key.toInt()]!!.add(Timestamp(this.updated, matchesPlayed, rankings, shots, goals,
                                    saves, assists,wins, mvps))
                        }
                    }
                }
            }
        } catch (error: JSONException) {
            android.util.Log.e("@playerUpdate","Parsing of the JSONResponse for updating the player failed with an JSONException " +error.message )
        }
    }
    private fun getRankings(processedSeason:JSONObject):Array<Ranking>{
        return arrayOf(Ranking(processedSeason.getJSONObject(""+10).getInt("rankPoints"),
                processedSeason.getJSONObject(""+10).getInt("tier"),
                processedSeason.getJSONObject(""+10).getInt("division")),
                Ranking(processedSeason.getJSONObject(""+11).getInt("rankPoints"),
                        processedSeason.getJSONObject(""+11).getInt("tier"),
                        processedSeason.getJSONObject(""+11).getInt("division")),
                Ranking(processedSeason.getJSONObject(""+12).getInt("rankPoints"),
                        processedSeason.getJSONObject(""+12).getInt("tier"),
                        processedSeason.getJSONObject(""+12).getInt("division")),
                Ranking(processedSeason.getJSONObject(""+13).getInt("rankPoints"),
                        processedSeason.getJSONObject(""+13).getInt("tier"),
                        processedSeason.getJSONObject(""+13).getInt("division")))

    }
}
